router/h2/src/main/scala/io/buoyant/router/h2/ClassifierFilter.scala

```diff
@@ -1,50 +1,155 @@
-  val role = Stack.Role("Classifier")
+  val role = Stack.Role("H2Classifier")
 
   val SuccessClassHeader = "l5d-success-class"
 
-  def module: Stackable[ServiceFactory[Request, Response]] =
-    new Stack.Module1[param.ResponseClassifier, ServiceFactory[Request, Response]] {
-      val role = ClassifierFilter.role
-      val description = "Sets the response classification into a header"
-      def make(classifierP: param.ResponseClassifier, next: ServiceFactory[Request, Response]) = {
-        val param.ResponseClassifier(classifier) = classifierP
+  protected val log = Logger.get("H2ClassifierFilter")
+
+  val module: Stackable[ServiceFactory[Request, Response]] =
+    new Stack.Module1[param.H2Classifier, ServiceFactory[Request, Response]] {
+      override val role: Stack.Role = ClassifierFilter.role
+      override val description = "Sets the stream classification into a header"
+      override def make(
+        classifierP: param.H2Classifier,
+        next: ServiceFactory[Request, Response]
+      ): ServiceFactory[Request, Response] = {
+        val param.H2Classifier(classifier) = classifierP
         new ClassifierFilter(classifier).andThen(next)
       }
     }
 
-  val successClassClassifier: ResponseClassifier = {
-    case ReqRep(req, Return(rep: Response)) if rep.headers.contains(SuccessClassHeader) =>
-      val success = rep.headers.get(SuccessClassHeader)
-        .flatMap { value =>
-          Try(value.toDouble).toOption
-        }.getOrElse(0.0)
-      if (success > 0.0)
-        ResponseClass.Successful(success)
-      else
-        ResponseClass.Failed(false)
+  /**
+   * Extractor for the `l5d-success-class` header field in streams/responses
+   */
+  private[this] object GetSuccessClass {
+    @inline def getSuccessClass(headers: Headers): Option[ResponseClass] =
+      headers.get(SuccessClassHeader).map { value =>
+        val success = Try { value.toDouble }.getOrElse {
+          log.warning(s"invalid `l5d-success-class` value $value, assumed failure")
+          0.0
+        }
+        if (success > 0.0) ResponseClass.Successful(success)
+        else ResponseClass.Failed(false)
+      }
+
+    @inline def unapply(frame: Frame): Option[ResponseClass] =
+      frame match {
+        case trailers: Trailers => getSuccessClass(trailers)
+        case _ => None
+      }
+
+    @inline def unapply(message: Message): Option[ResponseClass] =
+      getSuccessClass(message.headers)
+  }
+
+  /**
+   * An `H2Classifier` that classifies responses/streams based on the
+   * l5d-success-class` header.
+   *
+   * @note that this classifier is undefined for responses or streams
+   *       lacking this header.
+   */
+  object SuccessClassClassifier extends H2Classifier {
+    override val streamClassifier: PartialFunction[H2ReqRepFrame, ResponseClass] = {
+      case H2ReqRepFrame(_, Return((_, Some(Return(GetSuccessClass(c)))))) => c
+      case H2ReqRepFrame(_, Return((GetSuccessClass(c), Some(Return(_))))) => c
```

Comment:
I think this case should get caught in early classification (but I guess there's no harm in having it here too)